controller/object_store_controller.go

```diff
@@ -0,0 +1,1173 @@
+		"--rgw-dns-name",
+		fmt.Sprintf("%v.%v.svc", store.Name, osc.namespace),
+	}
+	for _, endpoint := range store.Spec.Endpoints {
+		domainNameArgs = append(domainNameArgs, "--rgw-dns-name")
+		domainNameArgs = append(domainNameArgs, endpoint.DomainName)
+	}
+
+	secret, err := osc.ds.GetSecret(osc.namespace, store.Spec.Credentials.Name)
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		return nil, errors.Wrapf(err, "failed to find secret %v", store.Spec.Credentials.Name)
+	}
+
+	env := []corev1.EnvFromSource{}
+	if secret != nil {
+		env = append(env, corev1.EnvFromSource{
+			SecretRef: &corev1.SecretEnvSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: store.Spec.Credentials.Name,
+				},
+			},
+		})
+	}
+
+	dpl := appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            store.Name,
+			Namespace:       osc.namespace,
+			Labels:          types.GetObjectStoreLabels(store),
+			OwnerReferences: osc.ds.GetOwnerReferencesForObjectStore(store),
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: osc.ds.GetObjectStoreSelectorLabels(store),
+			},
+			// an s3gw instance must have exclusive access to the volume, so we can
+			// only spawn one replica (i.e. one s3gw instance) per object-store.
+			// Due to the way the struct works, an allocated integer has to be used
+			// here and not a constant.
+			Replicas: int32Ptr(1),
+			Strategy: appsv1.DeploymentStrategy{
+				Type: appsv1.RecreateDeploymentStrategyType,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: osc.ds.GetObjectStoreSelectorLabels(store),
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  types.ObjectStoreContainerName,
+							Image: osc.getObjectStoreImage(store),
+							Args: append([]string{
+								"--rgw-backend-store", "sfs",
+								"--debug-rgw", fmt.Sprintf("%v", types.ObjectStoreLogLevel),
+								"--rgw_frontends", fmt.Sprintf(
+									"beast port=%d, status port=%d",
+									types.ObjectStoreContainerPort,
+									types.ObjectStoreStatusContainerPort,
+								),
+							}, domainNameArgs...),
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "s3",
+									ContainerPort: types.ObjectStoreContainerPort,
+									Protocol:      "TCP",
+								},
+								{
+									Name:          "status",
+									ContainerPort: types.ObjectStoreStatusContainerPort,
+									Protocol:      "TCP",
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "RGW_S3GW_ENABLE_TELEMETRY",
+									Value: "true",
+								},
+							},
+							EnvFrom: env,
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      genVolumeMountName(store),
+									MountPath: "/data",
+								},
+							},
+						},
+						{
+							Name:  types.ObjectStoreUIContainerName,
+							Image: osc.getObjectStoreUIImage(store),
```

Comment:
Since we already check the image inside the webhook, suggesting to set the image directly here and remove the function `getObjectStoreUIImage`
```suggestion
							Image: store.Spec.UIImage,
```